File: native/cpp/GFPCM.cpp
#include "GFPCM.h"
#include <iostream>
#include <limits> // For numeric_limits
#include <random> // For random_device, mt19937, uniform_int_distribution

GFPCM::GFPCM(int num_clusters, int max_iter, float m, float p, float w_prob, float noise)
	: BaseFCM(num_clusters, max_iter, m, noise), p(p), w_prob(w_prob)
{
	// Additional initialization specific to GFPCM if needed
}

void GFPCM::setParameters(const std::unordered_map<std::string, double> &params)
{
	BaseFCM::setParameters(params); // Call base class method
	if (params.find("p") != params.end())
	{
		p = static_cast<float>(params.at("p"));
	}
	if (params.find("w_prob") != params.end())
	{
		w_prob = static_cast<float>(params.at("w_prob"));
	}
}

std::vector<std::string> GFPCM::getParameterNames()
{
	auto base_params = BaseFCM::getParameterNames();
	base_params.push_back("p");
	base_params.push_back("w_prob");
	return base_params;
}

void GFPCM::fit(const Eigen::MatrixXd &X_in)
{
	if (!centroids_set)
	{
		throw std::runtime_error("Centroids must be set before calling fit.");
	}

	Eigen::MatrixXd X = getXWithNoise(X_in);

	int z = X.rows(); // Number of features (dimensions)
	int n = X.cols(); // Number of samples (data points)

	member.resize(num_clusters, n);
	Eigen::MatrixXd u = Eigen::MatrixXd::Zero(num_clusters, n);
	Eigen::MatrixXd t = Eigen::MatrixXd::Zero(num_clusters, n);

	float deriv_m = -2.0f / (m - 1.0f);
	float deriv_p = -2.0f / (p - 1.0f);

	for (int iter = 0; iter < max_iter; ++iter)
	{
		// Update u (probabilistic membership)
		for (int k = 0; k < n; ++k)
		{
			float szum = 0.0f;
			for (int i = 0; i < num_clusters; ++i)
			{
				float dist = (X.col(k) - centroids.col(i)).norm();
				if (dist < std::numeric_limits<float>::epsilon())
				{
					u(i, k) = std::numeric_limits<float>::infinity();
				}
				else
				{
					u(i, k) = std::pow(dist, deriv_m);
				}
				szum += u(i, k);
			}
			for (int i = 0; i < num_clusters; ++i)
			{
				if (szum > std::numeric_limits<float>::epsilon())
				{
					u(i, k) /= szum;
				}
				else
				{
					u(i, k) = 0.0f;
				}
			}
		}

		// Update t (possibilistic membership)
		for (int i = 0; i < num_clusters; ++i)
		{
			float szum = 0.0f;
			for (int k = 0; k < n; ++k)
			{
				float dist = (X.col(k) - centroids.col(i)).norm();
				if (dist < std::numeric_limits<float>::epsilon())
				{
					t(i, k) = std::numeric_limits<float>::infinity();
				}
				else
				{
					t(i, k) = std::pow(dist, deriv_p);
				}
				szum += t(i, k);
			}
			for (int k = 0; k < n; ++k)
			{
				if (szum > std::numeric_limits<float>::epsilon())
				{
					t(i, k) /= szum;
				}
				else
				{
					t(i, k) = 0.0f;
				}
			}
		}

		// Update centroids
		for (int i = 0; i < num_clusters; ++i)
		{
			Eigen::VectorXd sum_up = Eigen::VectorXd::Zero(z);
			float sum_dn = 0.0f;

			for (int k = 0; k < n; ++k)
			{
				float term_u = std::pow(u(i, k), m);
				float term_t = std::pow(t(i, k), p);
				float weight = term_u + w_prob * term_t;
				sum_up += weight * X.col(k);
				sum_dn += weight;
			}

			if (sum_dn > std::numeric_limits<float>::epsilon())
			{
				centroids.col(i) = sum_up / sum_dn;
			}
			else
			{
				// Keep previous centroid value if sum_dn is zero
			}
		}
	}

	// After iterations, set member matrix and trained flag
	// The Python implementation uses u ** self.m + self.w_prob * t ** self.p for _member.
	// For predicted labels, we'll use the 'u' matrix as it represents fuzzy membership.
	member = u; // We'll use 'u' as the primary membership for label prediction

	// Calculate predicted labels
	predictedLabels.clear();
	predictedLabels.reserve(n);
	for (int k = 0; k < n; ++k)
	{
		Eigen::MatrixXf::Index max_row;
		u.col(k).maxCoeff(&max_row);
		predictedLabels.push_back(static_cast<int>(max_row));
	}

	trained = true;
}

File: native/rust/algofuzz/src-nalgebra/src/gfpcm.rs
use nalgebra::{DMatrix, DVector};
use crate::centroid_strategy::{CentroidStrategy, CentroidUtility};

pub struct GFPCM {
    num_clusters: usize,
    max_iter: usize,
    m: f64,
    p: f64,
    w_prob: f64,
    centroids: DMatrix<f64>,
    centroid_strategy: CentroidStrategy,
    pub trained: bool,
    eta: DMatrix<f64>,
    pub member: DMatrix<f64>,
    pub predicted_labels: Vec<usize>,
}

impl GFPCM {
    pub fn new(num_clusters: usize, max_iter: usize, m: f64, p: f64, w_prob: f64, centroid_strategy: CentroidStrategy) -> GFPCM {
        GFPCM {
            num_clusters,
            max_iter,
            m,
            p,
            w_prob,
            centroids: DMatrix::zeros(1, 1),
            trained: false,
            eta: DMatrix::zeros(1, 1),
            member: DMatrix::zeros(1, 1),
            predicted_labels: Vec::new(),
            centroid_strategy,
        }
    }

    pub fn fit(&mut self, X: &DMatrix<f64>, true_labels: &Vec<usize>) {
        let actual_n = X.ncols();

        let z = X.nrows();
        let n = X.ncols();

        let mut u = DMatrix::zeros(self.num_clusters, n);
        let mut t = DMatrix::zeros(self.num_clusters, n);

        let deriv_m = -2.0 / (self.m - 1.0);
        let deriv_p = -2.0 / (self.p - 1.0);

        // Initialize centroids (assuming a function initialize_clusters exists)
        self.centroids = DMatrix::zeros(z, self.num_clusters);
        CentroidUtility::initialize_clusters(&X, &mut self.centroids, self.num_clusters, &self.centroid_strategy);

        for _ in 0..self.max_iter {
            // Update u
            for k in 0..n {
                let mut szum = 0.0;
                for i in 0..self.num_clusters {
                    u[(i, k)] = (X.column(k) - self.centroids.column(i))
                        .norm()
                        .powf(deriv_m);
                    szum += u[(i, k)];
                }
                for i in 0..self.num_clusters {
                    u[(i, k)] /= szum;
                }
            }

            // Update t
            for i in 0..self.num_clusters {
                let mut szum = 0.0;
                for k in 0..n {
                    t[(i, k)] = (X.column(k) - self.centroids.column(i))
                        .norm()
                        .powf(deriv_p);
                    szum += t[(i, k)];
                }
                for k in 0..n {
                    t[(i, k)] /= szum;
                }
            }

            // Update centroids
            for i in 0..self.num_clusters {
                let mut sumup = DVector::zeros(z);
                let mut sumdn = 0.0;
                for k in 0..n {
                    let weight = u[(i, k)].powf(self.m) + self.w_prob * t[(i, k)].powf(self.p);
                    sumup += weight * X.column(k);
                    sumdn += weight;
                }
                self.centroids.set_column(i, &(sumup / sumdn));
            }
        }

        self.eta = DMatrix::zeros(self.num_clusters, n);
        self.member = (u.map(|v| v.powf(self.m)) + t.map(|v| v.powf(self.p)) * self.w_prob).clone();
        self.trained = true;

        //println!("{:?}", self.member);

        self.predicted_labels = Vec::new();

        for k in 0..actual_n {
            let mut best_cluster = 0;
            let mut max_membership = 0.0;

            // Find the cluster with the highest membership value
            for i in 0..self.num_clusters {
                if self.member[(i, k)] > max_membership {
                    max_membership = self.member[(i, k)];
                    best_cluster = i;
                }
            }

            self.predicted_labels.push(best_cluster);
        }

        // Compute confusion matrix (assuming a function compute_confusion_matrix exists)
        let confusion_matrix = self.compute_confusion_matrix(true_labels, &self.predicted_labels);
        //println!("Confusion matrix: {:?}", confusion_matrix);
    }

    fn compute_confusion_matrix(
        &self,
        _true_labels: &Vec<usize>,
        _predicted_labels: &Vec<usize>,
    ) -> DMatrix<f64> {
        // Implementation for computing confusion matrix
        DMatrix::zeros(1, 1)
    }
}

/*
fn main() {
    let num_clusters = 3;
    let max_iter = 100;
    let m = 2.0;
    let p = 2.0;
    let w_prob = 0.5;
    let noise = 0.0;

    let X = DMatrix::from_row_slice(4, 4, &[
        1.0, 2.0, 3.0, 4.0,
        5.0, 6.0, 7.0, 8.0,
        9.0, 10.0, 11.0, 12.0,
        13.0, 14.0, 15.0, 16.0,
    ]);

    let true_labels = vec![0, 1, 2, 0];

    let mut gfpcm = GFPCM {
        num_clusters,
        max_iter,
        m,
        p,
        w_prob,
        noise,
        centroids: DMatrix::zeros(4, num_clusters),
        trained: false,
        eta: DMatrix::zeros(num_clusters, X.ncols()),
        member: DMatrix::zeros(num_clusters, X.ncols()),
    };

    gfpcm.fit(&X, &true_labels);
}*/